en/deploy-br-federation-manager.md

```diff
@@ -0,0 +1,350 @@
+
+{{< copyable "shell-regular" >}}
+
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: br-federation-member
+  namespace: tidb-admin
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: br-federation-manager:br-federation-member
+rules:
+- apiGroups:
+  - pingcap.com
+  resources:
+  - backups
+  - restores
+  verbs:
+  - '*'
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: br-federation-manager:br-federation-member
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: br-federation-manager:br-federation-member
+subjects:
+- kind: ServiceAccount
+  name: br-federation-member
+  namespace: tidb-admin
+```
+
+#### Generate Kubeconfig
+
+{{< copyable "shell-regular" >}}
+
+```shell
+export USER_TOKEN_NAME=$(kubectl -n tidb-admin get serviceaccount br-federation-member -o=jsonpath='{.secrets[0].name}')
+export USER_TOKEN_VALUE=$(kubectl -n tidb-admin get secret/${USER_TOKEN_NAME} -o=go-template='{{.data.token}}' | base64 --decode)
+export CURRENT_CONTEXT=$(kubectl config current-context)
+export CURRENT_CLUSTER=$(kubectl config view --raw -o=go-template='{{range .contexts}}{{if eq .name "'''${CURRENT_CONTEXT}'''"}}{{ index .context "cluster" }}{{end}}{{end}}')
+export CLUSTER_CA=$(kubectl config view --raw -o=go-template='{{range .clusters}}{{if eq .name "'''${CURRENT_CLUSTER}'''"}}"{{with index .cluster "certificate-authority-data" }}{{.}}{{end}}"{{ end }}{{ end }}')
+export CLUSTER_SERVER=$(kubectl config view --raw -o=go-template='{{range .clusters}}{{if eq .name "'''${CURRENT_CLUSTER}'''"}}{{ .cluster.server }}{{end}}{{ end }}')
+
+cat << EOF > {k8s-name}-kubeconfig
+apiVersion: v1
+kind: Config
+current-context: ${CURRENT_CONTEXT}
+contexts:
+- name: ${CURRENT_CONTEXT}
+  context:
+    cluster: ${CURRENT_CONTEXT}
+    user: br-federation-member
+    namespace: kube-system
+clusters:
+- name: ${CURRENT_CONTEXT}
+  cluster:
+    certificate-authority-data: ${CLUSTER_CA}
+    server: ${CLUSTER_SERVER}
+users:
+- name: br-federation-member
+  user:
+    token: ${USER_TOKEN_VALUE}
+EOF
+```
+
+### Control Plane
+
+You should select one Kubernetes cluster as control plane to deploy BR Federation Manager. You need only execute steps below in the control plane.
+
+#### Create CRD
+
+{{< copyable "shell-regular" >}}
+
+```shell
+kubectl create -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/federation-crd.yaml
+```
+
+#### Create Secret
+
+You already have kubeconfig files of data planes. Now, you need encode all the kubeconfig files to a secret.
+
+Firstly, encode every kubeconfig file by `base64 -i {k8s-name}-kubeconfig`.
+
+Secondly, put all the output of first step to a secret object.
+
+{{< copyable "shell-regular" >}}
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: br-federation-kubeconfig
+type: Opaque
+data:
+  {k8s-name1}: {encoded-kubeconfig-1}
```

Comment:
Instead of having a mapping from k8s-name to encoded kubeconfig, we can instead provide single kubeconfig that has multiple contexts, one for each of the k8s-name1, k8s-name2, etc.

Kubeconfigs already supports multiple contexts, so we don't need to invent a new format here map from clusters to kubeconfigs.
```
apiVersion: v1
kind: Config
clusters:
- cluster:
    server: https://k8s-name1
  name: k8s-name1
- cluster:
    server: https://k8s-name2
  name: k8s-name2
contexts:
- context:
    cluster:  k8s-name1
    user:  user
  name: k8s-name1
- context:
    cluster:  k8s-name2
    user:  user
  name: k8s-name2
preferences: {}
users:
- name:  user
  user:
    token: <secret-token-here>
```

From my brief review of the code, the main change is that https://github.com/pingcap/tidb-operator/blob/8b1aa3bcd3d3aacbfca04c31cbe7a9231758c2c7/pkg/controller/br_fed_config.go#L76 command line arg should instead point to a single kubeconfig with multiple contexts

Reply:
It's a good suggestion, we can change to use one kubeconfig and change the command line arg to a file path. But I think we still need a secret to store the kubeconfig so that we can mount the kubeconfig to the container of `tidb-operator` directly.

Reply:
I have created an [issue](https://github.com/pingcap/tidb-operator/issues/5183) to track it.